### PR TITLE
Prevent resourceVersion updates for custom resources on no-op writes

### PR DESCRIPTION
**Pull request by nikhita**
Fixes partly https://github.com/kubernetes/kubernetes/issues/67541

For ObjectMeta pruning, we round trip by marshalling and unmarshalling. If the ObjectMeta contained any strings with `""` (or other fields with empty values)  _and_ the respective fields are `omitempty`, those fields will be lost in the round trip process.

This makes ObjectMeta after the no-op write different from the one before the write.

Resource version is incremented every time data is written to etcd. Writes to etcd short-circuit if the bytes being written are identical to the bytes already present.

So this ends up incrementing the `resourceVersion` even on no-op writes. This PR updates the `BeforeUpdate` function such that omitempty fields have values set only if they are non-zero so that they produce an unstructured object that matches ObjectMeta omitempty semantics.

/sig api-machinery
/kind bug
/area custom-resources
/assign sttts liggitt 

**Release note**:

```release-note
Prevent `resourceVersion` updates for custom resources on no-op writes.
```

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 1]
```diff
@@ -893,3 +886,101 @@ func setObjectMeta(u *unstructured.Unstructured, objectMeta *metav1.ObjectMeta)
...
 	return nil
 }
+
+type implicitField struct {
+	path   string
+	exists bool
+	value  interface{}
+}
+
+func getImplicitFields(u *unstructured.Unstructured) ([]implicitField, error) {
```
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: I want to get an initial review on this before adding all fields in objectMeta here.

Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: And if we end up setting these fields explicitly, do we still want to round trip through marshalling and unmarshalling or go field by field accumulating into the metadata object by iterating through the `metadataMap`?

Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: ref https://github.com/kubernetes/kubernetes/pull/64267#discussion_r190924277

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: I don't think this is the right layer at which to do this. I think the unstructured objectmeta accessor implementations should be adjusted to produce an unstructured object that matches ObjectMeta omitempty semantics

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: For example, if SetClusterName("") is called, that should remove the clusterName element from the unstructured object

Comment (yue9944882) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: tricky one: do we need to do cascading element removal in accessor for omitempty semantics?

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: I think all the accessor methods are for top level fields in metadata, so I'm not sure that's an issue

Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: updated.

Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: > I think the unstructured objectmeta accessor implementations should be adjusted to produce an unstructured object that matches ObjectMeta omitempty semantics

The accessors were already modified to support this for some of the fields. Added for the rest.

**Comment by liggitt** [comment]
> I'm wondering if this is fine to cherry-pick to 1.11 or not. If someone uses client-go to interact with unstructured objects, this could lead to breaking changes for them.

It's important to resolve the incrementing resourceVersion issue for 1.11. Was the only field force-set to an empty value in BeforeCreate the clusterName and initializers fields? We may be able to make a more targeted fix for 1.11 to only force-set those fields to empty values if they contain non-empty values

**Comment by nikhita** [comment]
> We may be able to make a more targeted fix for 1.11 to only force-set those fields to empty values if they contain non-empty values

:+1: 

> Were the only fields force-set to an empty value in BeforeCreate the clusterName and initializers fields?

In [BeforeCreate](https://github.com/kubernetes/kubernetes/blob/31420467ae9503850515786660fa3e93e9d140b2/staging/src/k8s.io/apiserver/pkg/registry/rest/create.go#L74):

1. Namespace
2. DeletionTimestamp
3. DeletionGracePeriodSeconds
4. Initializers
5. ClusterName

In [BeforeUpdate](https://github.com/kubernetes/kubernetes/blob/31420467ae9503850515786660fa3e93e9d140b2/staging/src/k8s.io/apiserver/pkg/registry/rest/update.go#L86):

1. Namespace
2. Initializers
3. ClusterName

But I've noticed that we need to include selfLink and resourceVersion as well (they were creating problems in tests). I think this is because they are cleared prior to writing to etcd.

**Comment by nikhita** [comment]
I'll create a PR for 1.11 once this looks good to merge.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go [hunk 2]
```diff
@@ -179,6 +179,11 @@ func (u *Unstructured) GetOwnerReferences() []metav1.OwnerReference {
 }
 
 func (u *Unstructured) SetOwnerReferences(references []metav1.OwnerReference) {
+	if references == nil {
+		RemoveNestedField(u.Object, "metadata", "ownerReferences")
```
Comment (yue9944882) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go]: hi @nikhita, how about having an additional check to respect the field tag `omitempty` via `reflect`? we might only do the removal when it's tagged, amiright?

Comment (yue9944882) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go]: it's working tho, but a dynamic approach would be better for maintenance

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go]: Reflection is slow to use in the normal path. A reflection based unit test to catch drift might be better.

Comment (yue9944882) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go]: > Reflection is slow to use in the normal path.

true. it's a heavy cost to do the reflection every time accessing normal paths. or we could simply have a thread-unsafe map like `map[(path)string](isOmitemptyField)bool` and put all things we need inside `init()` call

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go [hunk 3]
```diff
@@ -30,3 +33,43 @@ func TestNilUnstructuredContent(t *testing.T) {
...
+	var u Unstructured
+
+	// start with an empty metadata
+	emptyMetadata := make(map[string]interface{})
+	u.setNestedField(emptyMetadata, "metadata")
+
+	// set zero values for all fields in metadata explicitly
+	// since all fields in objectMeta are omitempty, these fields should never be set
+	// and the resulting metadata should be an empty metadata
+	u.SetName("")
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go]: shouldn't we set random values first?

Comment (sttts) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go]: I meant to do fuzzing before these to catch missing calls. If a field is added to ObjectMeta now, this list gets incomplete, but we don't notice. If we fuzz before, we would.

File: staging/src/k8s.io/apimachinery/pkg/runtime/converter.go [hunk 4]
```diff
@@ -790,6 +790,9 @@ func structToUnstructured(sv, dv reflect.Value) error {
 			if err := toUnstructured(fv, subv); err != nil {
 				return err
 			}
+			if fieldInfo.omitempty && subv.IsNil() {
+				continue
```
[the remark this answers is not in the document]
Comment (nikhita) [reply staging/src/k8s.io/apimachinery/pkg/runtime/converter.go]: Looks like this isnt' right. json.Marshal +Unmarshal will set `creationTimestamp: nil` and we always need to ensure that it satisfies json marshalling + unmarshalling.

https://github.com/kubernetes/kubernetes/blob/39e341c11f9a18d7fcbf78e4dce6801131e2ba38/staging/src/k8s.io/apimachinery/pkg/runtime/converter.go#L425

This will fail the already existing round trip test for unstructured:

https://github.com/kubernetes/kubernetes/blob/39e341c11f9a18d7fcbf78e4dce6801131e2ba38/pkg/api/testing/unstructured_test.go#L120

Interesting to note that we also have a round trip test specifically for creationTimestamp but it just checks if conversion is possible and not that the round trip values matches.

https://github.com/kubernetes/kubernetes/blob/39e341c11f9a18d7fcbf78e4dce6801131e2ba38/pkg/api/testing/unstructured_test.go#L137

[the remark this answers is not in the document]
Comment (nikhita) [reply staging/src/k8s.io/apimachinery/pkg/runtime/converter.go]: What should be the intended behaviour here? Should `creationTimestamp` be present in the map or not? :thinking:

[the remark this answers is not in the document]
Comment (nikhita) [reply staging/src/k8s.io/apimachinery/pkg/runtime/converter.go]: > What should be the intended behaviour here? Should creationTimestamp be present in the map or not? thinking

answering myself: from the docs

> The "omitempty" option specifies that the field should be omitted from the encoding if the field has an empty value, defined as false, 0, a nil pointer, a nil interface value, and any empty array, slice, map, or string.

`omitempty` has no meaning for a struct, so `creationTimestamp` should be present in the map.  It is weird that we set the value as nil but in any case, it should be present in the unstructured object/map.

**Comment by nikhita** [comment]
Updated. Ptal.

File: staging/src/k8s.io/apimachinery/pkg/runtime/converter.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (yue9944882) [reply staging/src/k8s.io/apimachinery/pkg/runtime/converter.go]: shouldn't `omitempty` be removed from `CreateTimestamp` then? maybe via another pull?

[the remark this answers is not in the document]
Comment (nikhita) [reply staging/src/k8s.io/apimachinery/pkg/runtime/converter.go]: > shouldn't omitempty be removed from CreateTimestamp then? maybe in another pull? it

yeah, I think we should remove it too. But I thought it was suited for another PR (since it would involved schema changes too afaik and would be a breaking change).

Also, wanted to get feedback on the current changes before removing it. :)

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go [hunk 5]
```diff
@@ -14,19 +14,146 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package unstructured
+package unstructured_test
```
Comment (nikhita) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go]: renamed to fix import cycle:

```
import cycle not allowed in test
package k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured (test)
	imports k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/api/apitesting/fuzzer
	imports k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer
	imports k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/versioning
	imports k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured
```

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/basic_test.go [hunk 6]
```diff
@@ -678,18 +678,27 @@ func TestPatch(t *testing.T) {
...
+	patchedNoxuInstance, err = noxuNamespacedResourceClient.Patch("foo", types.MergePatchType, []byte(`{}`), metav1.UpdateOptions{})
 	if err != nil {
 		t.Fatalf("unexpected error: %v", err)
 	}
+	// an empty patch is a no-op patch. make sure it does not increment resourceVersion
+	if patchedNoxuInstance.GetResourceVersion() != resourceVersion {
+		t.Fatalf("unexpected error: resourceVersion changed on empty patch. expected %s, got %s", resourceVersion, patchedNoxuInstance.GetResourceVersion())
+	}
 
 	originalJSON, err := runtime.Encode(unstructured.UnstructuredJSONScheme, createdNoxuInstance)
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/basic_test.go]: this is referencing `createdNoxuInstance` which is no longer the result of the operation on line 689, right? should it change to `patchedNoxuInstance`?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go [hunk 7]
```diff
@@ -282,9 +319,14 @@ func (u *Unstructured) GetCreationTimestamp() metav1.Time {
 }
 
 func (u *Unstructured) SetCreationTimestamp(timestamp metav1.Time) {
+	// creationTimestamp is a struct (metav1.Time) so omitempty holds no meaning for it.
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go]: if this isn't causing issues with etcd resourceVersion changes (which it shouldn't, since we don't persist objects with zero-value creationTimestamps to etcd), let's not make this change in this PR. I'd actually like to see if it would be possible to go the other direction in a separate PR, and make ObjectMeta actually omit creationTimestamp when empty.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/basic_test.go (hunk 6 above)
Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/basic_test.go]: fixed this.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go (hunk 7 above)
Comment (nikhita) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go]: > if this isn't causing issues with etcd resourceVersion changes 

it doesn't :+1: 

> make ObjectMeta actually omit creationTimestamp when empty.

My understanding is that `omitEmpty` does not apply to structs and no matter what the value of `creationTimestamp` is, this should persist in the object.

**Comment by liggitt** [comment]
would it make sense to do the minimal fix with conditional zero-value sets in BeforeCreate/BeforeUpdate, along with the custom resource resourceVersion tests from this PR, so we can make sure that fixes the issue in a way we can safely pick to 1.11, then do these unstructured changes in a follow up?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go (hunk 7 above)
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go]: > My understanding is that omitEmpty does not apply to structs and no matter what the value of creationTimestamp is, this should persist in the object.

You are correct. I'd like to explore switching the creationTimestamp to a `*time.Time` so it can actually be omitted (verifying wire compatibility for existing json and proto, obviously)

**Comment by nikhita** [comment]
> would it make sense to do the minimal fix with conditional zero-value sets in BeforeCreate/BeforeUpdate, along with the custom resource resourceVersion tests from this PR, so we can make sure that fixes the issue in a way we can safely pick to 1.11, then do these unstructured changes in a follow up? I'd like the 1.11 pick to be an actual pick of a master branch change if possible.

Done. Ptal.

File: staging/src/k8s.io/apiserver/pkg/registry/rest/update.go [hunk 8]
```diff
@@ -83,35 +83,43 @@ func validateCommonFields(obj, old runtime.Object, strategy RESTUpdateStrategy)
...
+	if err != nil {
+		return err
+	}
+
 	if strategy.NamespaceScoped() {
 		if !ValidNamespace(ctx, objectMeta) {
 			return errors.NewBadRequest("the namespace of the provided object does not match the namespace sent on the request")
 		}
-	} else {
+	} else if len(oldMeta.GetNamespace()) == 0 {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/registry/rest/update.go]: `if len(objectMeta.GetNamespace()) > 0`

File: staging/src/k8s.io/apiserver/pkg/registry/rest/update.go [hunk 9]
```diff
@@ -83,35 +83,43 @@ func validateCommonFields(obj, old runtime.Object, strategy RESTUpdateStrategy)
...
 	// Ensure requests cannot update generation
-	oldMeta, err := meta.Accessor(old)
-	if err != nil {
-		return err
-	}
 	objectMeta.SetGeneration(oldMeta.GetGeneration())
 
 	// Ensure Initializers are not set unless the feature is enabled
 	if !utilfeature.DefaultFeatureGate.Enabled(features.Initializers) {
+		if oldMeta.GetInitializers() == nil {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/registry/rest/update.go]: unstructured accessor actually does the right thing with initializers already, so can drop this change

File: staging/src/k8s.io/apiserver/pkg/registry/rest/update.go [hunk 10]
```diff
@@ -83,35 +83,43 @@ func validateCommonFields(obj, old runtime.Object, strategy RESTUpdateStrategy)
...
+		}
 		oldMeta.SetInitializers(nil)
-		objectMeta.SetInitializers(nil)
 	}
 
 	strategy.PrepareForUpdate(ctx, obj, old)
 
 	// ClusterName is ignored and should not be saved
-	objectMeta.SetClusterName("")
+	if len(oldMeta.GetClusterName()) == 0 {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/registry/rest/update.go]: `if len(objectMeta.GetClusterName()) > 0`

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go (hunk 7 above)
Comment (nikhita) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go]: > I'd like to explore switching the creationTimestamp to a *time.Time so it can actually be omitted (verifying wire compatibility for existing json and proto, obviously)

created https://github.com/kubernetes/kubernetes/issues/67610 to track this.

**Comment by liggitt** [comment]
thanks, can you pick this to 1.11 once it merges?

**Comment by nikhita** [comment]
> thanks, can you pick this to 1.11 once it merges?

will do. also, I'll create a PR against master with the remaining changes.

**Comment by liggitt** [comment]
/hold

actually, I think the same change needs to be made in BeforeCreate()

File: staging/src/k8s.io/apiserver/pkg/registry/rest/meta.go [hunk 11]
```diff
@@ -28,7 +28,10 @@ import (
 func FillObjectMetaSystemFields(meta metav1.Object) {
 	meta.SetCreationTimestamp(metav1.Now())
 	meta.SetUID(uuid.NewUUID())
-	meta.SetSelfLink("")
+
+	if len(meta.GetSelfLink()) > 0 {
+		meta.SetSelfLink("")
```
Comment (nikhita) [inline staging/src/k8s.io/apiserver/pkg/registry/rest/meta.go]: @liggitt ptal if this is need.

I don't think this hurts though.

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/registry/rest/meta.go]: shouldn't be needed (PrepareObjectForStorage will do the same thing unconditionally). let's keep this PR scoped just to the things needed to fix the resourceVersion issue. fixing the unstructured accessor in the follow-up PR would mean we wouldn't care about this

File: staging/src/k8s.io/apiserver/pkg/registry/rest/create.go [hunk 12]
```diff
@@ -81,11 +81,17 @@ func BeforeCreate(strategy RESTCreateStrategy, ctx context.Context, obj runtime.
 		if !ValidNamespace(ctx, objectMeta) {
 			return errors.NewBadRequest("the namespace of the provided object does not match the namespace sent on the request")
 		}
-	} else {
+	} else if len(objectMeta.GetNamespace()) > 0 {
 		objectMeta.SetNamespace(metav1.NamespaceNone)
 	}
-	objectMeta.SetDeletionTimestamp(nil)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/registry/rest/create.go]: can drop the deletion changes... the unstructured accessor already handles those correctly

**Comment by nikhita** [comment]
For the record:

- https://github.com/kubernetes/kubernetes/pull/67616: cherry-pick of this PR to 1.11
- https://github.com/kubernetes/kubernetes/pull/67635: follow-up PR against master containing fixes for unstructured accessors

**Comment by nikhita** [comment]
Picked the test changes from @sttts's PR: https://github.com/kubernetes/kubernetes/pull/67637 (as discussed with @sttts on slack).
